File: src/cmcandy/java_language_answers/_0110.java
package cmcandy.java_language_answers;

/**
 * Created by CHEN on 2020/8/10.
 */
public class _0110 {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode(int x) { val = x; }
     * }
     */
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }
    static class Solution {
        public boolean isBalanced(TreeNode root) {
            if(root==null) return true;
            int left =  countChild(root.left);
            if(left==-1) return false;
            int right =  countChild(root.right);
            return right==-1||Math.abs(left-right)>1?false:true;
        }


        public int countChild(TreeNode root){
            if(root==null) return 0;
            else{
                int left = countChild(root.left);
                if(left==-1) return -1;
                int right = countChild(root.right);
                return Math.abs(left-right)>1||right==-1?-1:Math.max(left,right)+1;
            }
        }
    }
}

File: src/cmcandy/C_Language_Answers/_0110.cpp
#include <iostream>
#include <math.h>

using namespace ::std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution
{
public:
    bool isBalanced(TreeNode *root)
    {
        return funDepth(root) == -1 ? false : true;
    }

    int funDepth(TreeNode *root)
    {
        if (root == NULL)
            return 0;
        if (root->left == NULL && root->right == NULL)
            return 1;
        else
        {
            int depthl = funDepth(root->left);
            int depthr = funDepth(root->right);

            //如果平衡，则返回深度，否则返回-1
            if (depthr != -1 && depthl != -1 && abs(depthr - depthl) <= 1)
                return max(depthr, depthl) + 1;
            else
                return -1;
        }
    }
};